### PR TITLE
Return errors from RecordBatchReader

### LINKED ISSUE #4282: Python API: errors swallowed in RecordBatchReader
### What happens?

When query results are fetched via a RecordBatchReader and the query fails, `StopIteration` is raised and the actual DuckDB error message doesn't seem to be recoverable (trying to read another batch results in `OSError: Query Failed`). I'll open a PR with a proposed fix.

### To Reproduce

```
conn = duckdb.connect()
conn.execute("CREATE table t as select 'foo' as a;")
# try failing cast, should report corresponding error
conn.execute("SELECT cast(a as double) FROM t").fetch_record_batch(1024).read_next_batch()
```

### OS:

Windows 10, Linux

### DuckDB Version:

0.4.1-dev1101

### DuckDB Client:

Python

### Full Name:

In profile

### Affiliation:

In profile

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by eeroel**
Attempt to fix https://github.com/duckdb/duckdb/issues/4282
- Add a success check after `result.Fetch()` so that errors are detected
- Use `QueryResult.GetError()` to set last error for the arrow stream

### CODE REVIEW

File: tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py [hunk 1]
```diff
@@ -106,3 +106,15 @@ def test_record_coverage(self, duckdb_cursor):
...
+            return
+        duckdb_cursor = duckdb.connect()
+        duckdb_cursor.execute("CREATE table t as select 'foo' as a;")
+        query = duckdb_cursor.execute("SELECT cast(a as double) FROM t")
+        record_batch_reader = query.fetch_record_batch(1024)
+        try:
+            chunk = record_batch_reader.read_next_batch()
+            assert(False)
+        except Exception as e:
+            assert('Conversion Error' in str(e))
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py:120]:
```suggestion
        with pytest.raises(RuntimeError, match='Conversion Error'):
            record_batch_reader.read_next_batch()
```

File: tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py [hunk 2]
```diff
@@ -106,3 +106,15 @@ def test_record_coverage(self, duckdb_cursor):
         chunk = record_batch_reader.read_all()
         assert(len(chunk) == 2048)
 
+    def test_record_batch_query_error(self, duckdb_cursor):
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py:109]:
```suggestion
    def test_record_batch_query_error(self):
```

**Review by Mause — CHANGES_REQUESTED** [review]

**Comment by pdet** [comment]
@eeroel Many thanks for the PR! I think that after you apply @Mause 's suggested changes will be ready to be merged.

File: tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py (hunk 1 above)
Comment (eeroel) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py]: Not sure where it arises but in this case the exception is of type `OSError`. Test should be fixed now.

File: tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py [hunk 3]
```diff
@@ -106,3 +106,12 @@ def test_record_coverage(self, duckdb_cursor):
...
         assert(len(chunk) == 2048)
 
+    def test_record_batch_query_error(self):
+        if not can_run:
+            return
+        duckdb_cursor = duckdb.connect()
+        duckdb_cursor.execute("CREATE table t as select 'foo' as a;")
+        query = duckdb_cursor.execute("SELECT cast(a as double) FROM t")
+        record_batch_reader = query.fetch_record_batch(1024)
+        with pytest.raises(OSError, match='Conversion Error'):
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py:116]: Is this really throwing an `OSError`?

Comment (eeroel) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py:116]: Yes, it's raised in `pyarrow`, for example:
```Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow\ipc.pxi", line 682, in pyarrow.lib.RecordBatchReader.read_next_batch
  File "pyarrow\error.pxi", line 115, in pyarrow.lib.check_status
OSError: Query Failed```

Comment (Mause) [reply tools/pythonpkg/tests/fast/arrow/test_arrow_fetch_recordbatch.py:116]: Huh, interesting. Happy to merge this, though I might have to do some investigation to see if that exception type can be improved

**Review by Mause — APPROVED** [review]

**Pull request merged**